File: sol/729/729.cpp
#include <set>

struct MyCalendar {
    set<pair<int, int>> data;

    MyCalendar(): data() {}
    
    bool book(int start, int end) {
        auto it = this->data.lower_bound({start, end});
        if ((it != this->data.begin() && std::prev(it)->second > start)
                || (it != this->data.end() && it->first < end)){
            return false;
        }
        this->data.insert({start, end});
        return true;
    }
};

File: sol/729/729.py
from sortedcontainers import SortedDict

class MyCalendar:
    def __init__(self):
        self.data = SortedDict()

    def book(self, start: int, end: int) -> bool:
        idx = self.data.bisect_right(start)
        if idx != len(self.data) and self.data.peekitem(idx)[1] < end:
            return False
        self.data[end] = start
        return True

